### PR TITLE
ARROW-11984: [C++][Gandiva] Implement SHA1 and SHA256 functions

### PR DESCRIPTION
**Pull request by jpedroantunes**
Implement SHA1 and SHA256 functions on Gandiva module.
Used OpenSSL to run the SHA algorithm for numeric and string values.

JIRA issue: https://issues.apache.org/jira/browse/ARROW-11984

### CODE REVIEW

File: cpp/src/gandiva/function_registry_common.h [hunk 1]
```diff
@@ -199,6 +199,28 @@ typedef std::unordered_map<const FunctionSignature*, const NativeFunction*, KeyH
                  DataTypeVector{TYPE(), int64()}, int64(), kResultNullNever, \
                  ARROW_STRINGIFY(NAME##WithSeed_##TYPE))
 
+// HashSHA128 functions that :
+// - NULL handling is of type NULL_NEVER
+// - can return errors
+//
+// The function name includes the base name & input type name. gdv_fn_sha128_float64
+#define HASH_SHA128_NULL_NEVER(NAME, ALIASES, TYPE)                           \
```
Comment (projjal) [inline cpp/src/gandiva/function_registry_common.h:207]: I think it is called SHA1 and not SHA128. The digest bit size is also 160 and not 128

File: cpp/src/gandiva/hash_utils.cc [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  const char* HashUtils::GetHash(int64_t context,
+                                 const void* message,
+                                 size_t message_length,
+                                 const EVP_MD *hash_type,
+                                 int result_buf_size,
+                                 int32_t *out_length) {
+    EVP_MD_CTX *md_ctx = EVP_MD_CTX_new();
+
+    if (md_ctx == nullptr) {
+      HashUtils::ErrorMessage(context, "Could not allocate memory "
```
Comment (projjal) [inline cpp/src/gandiva/hash_utils.cc:53]: -> "Could not create context"

File: cpp/src/gandiva/hash_utils.cc [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+    if (result == nullptr) {
+      HashUtils::ErrorMessage(context, "Could not allocate memory "
+									   "for SHA processing.");
+      EVP_MD_CTX_free(md_ctx);
+      return "";
+    }
+
+    unsigned int result_length;
+    EVP_DigestFinal_ex(md_ctx, result, &result_length);
```
Comment (projjal) [inline cpp/src/gandiva/hash_utils.cc:85]: add a check result_length == hash_size
and result_buf_size == 2 * hash_size

File: cpp/src/gandiva/hash_utils.cc [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    // Add the NULL character to shows the end of the string
+    result_buffer[result_buf_size - 1] = '\0';
+
+    // free the resources to avoid memory leaks
+    EVP_MD_CTX_free(md_ctx);
+    free(result);
+
+    *out_length = strlen(result_buffer);
```
Comment (projjal) [inline cpp/src/gandiva/hash_utils.cc:118]: *out_length = result_buf_size

File: cpp/src/gandiva/hash_utils.h [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+#include <cstdlib>
+#include <cstdint>
+#include "gandiva/visibility.h"
+#include "openssl/evp.h"
+
+namespace gandiva {
+class GANDIVA_EXPORT HashUtils {
+ public:
+  static const char *HashUsingSha256(int64_t context,
```
Comment (emkornfield) [inline cpp/src/gandiva/hash_utils.h:29]: documentations for these methods?  The pattern here seems strange to have all static methods on the class why not just make these first class functions without the class?  At least a comment on that aspect would be worthwhile also.

File: cpp/src/gandiva/tests/hash_test.cc [hunk 6]
```diff
@@ -147,4 +153,252 @@ TEST_F(TestHash, TestBuf) {
...
+            MakeArrowArrayUtf8({"foo", first_string, second_string},
+							   {false, true, true});
+
+    // prepare input record batch
+    auto in_batch = arrow::RecordBatch::Make(schema, num_records, {array_a});
+
+    // Evaluate expression
+    arrow::ArrayVector outputs;
+    status = projector->Evaluate(*in_batch, pool_, &outputs);
+    EXPECT_TRUE(status.ok());
```
Comment (emkornfield) [inline cpp/src/gandiva/tests/hash_test.cc:393]: I believe in Arrow we have ASSERT_OK, which should porbably be used (I don't think you want to continue if status isn't OK?

File: cpp/src/gandiva/hash_utils.cc [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+#include "gandiva/execution_context.h"
+#include "gandiva/gdv_function_stubs.h"
+
+namespace gandiva {
+  const char* HashUtils::HashUsingSha256(int64_t context,
+                                         const void* message,
+                                         size_t message_length,
+                                         int32_t *out_length) {
+    // The buffer size is the hash size + null character
+    int sha256_result_length = 65;
```
Comment (emkornfield) [inline cpp/src/gandiva/hash_utils.cc:30]: constexpr?

File: cpp/src/gandiva/hash_utils.cc [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+    return HashUtils::GetHash(context, message, message_length, EVP_sha256(),
+							  sha256_result_length, out_length);
+  }
+  const char* HashUtils::HashUsingSha128(int64_t context,
+                                         const void* message,
+                                         size_t message_length,
+                                         int32_t *out_length) {
+    // The buffer size is the hash size + null character
+    int sha128_result_length = 41;
```
Comment (emkornfield) [inline cpp/src/gandiva/hash_utils.cc:39]: constexpr?

File: cpp/src/gandiva/hash_utils.cc [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    if (hex_buffer == nullptr || result_buffer == nullptr) {
+      gdv_fn_context_set_error_msg(context, "Could not allocate memory "
+                                       "for the result buffers.");
+      EVP_MD_CTX_free(md_ctx);
+      return "";
+    }
+
+    for (unsigned int j = 0; j < result_length; j++) {
+      unsigned char hex_number = result[j];
+      snprintf(hex_buffer, tmp_buf_len, "%02x", hex_number);
```
Comment (emkornfield) [inline cpp/src/gandiva/hash_utils.cc:107]: it seems like you should snprintf to the final result_buffer directly and update the pointer using the returned value? (you porbably also want to at least DCHECK the result.

File: cpp/src/gandiva/hash_utils.cc [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+      HashUtils::ErrorMessage(context, "Could not allocate memory "
+									   "for SHA processing.");
+      return "";
+    }
+
+    int evp_success_status = 1;
+
+    if (EVP_DigestInit_ex(md_ctx, hash_type, nullptr) != evp_success_status) {
+      HashUtils::ErrorMessage(context, "Could not obtain the hash "
+									   "for the defined value.");
```
Comment (emkornfield) [inline cpp/src/gandiva/hash_utils.cc:62]: this formatting looks strange.

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 11]
```diff
@@ -122,6 +123,133 @@ int32_t gdv_fn_populate_varlen_vector(int64_t context_ptr, int8_t* data_ptr,
...
+  GANDIVA_EXPORT                                                                        \
+  const char *gdv_fn_sha256_##TYPE(int64_t context, gdv_##TYPE value,                   \
+                                        bool validity, int32_t *out_length) {           \
+    if (!validity) {                                                                    \
+      *out_length = 0;                                                                  \
+      return "";                                                                        \
+    }                                                                                   \
+    auto value_as_long = gandiva::HashUtils::DoubleToLong((double)value);               \
+    const char *result = gandiva::HashUtils::HashUsingSha256(context,                   \
+                                                               &value_as_long,          \
```
Comment (emkornfield) [inline cpp/src/gandiva/gdv_function_stubs.cc:168]: it looks like this is a void*?  Why is the doubletolong conversion necessary?

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 12]
```diff
@@ -122,6 +123,133 @@ int32_t gdv_fn_populate_varlen_vector(int64_t context_ptr, int8_t* data_ptr,
...
+SHA_NUMERIC_BOOL_DATE_PARAMS(SHA128_HASH_FUNCTION)
+SHA_VAR_LEN_PARAMS(SHA128_HASH_FUNCTION_BUF)
+
+#undef SHA_NUMERIC_BOOL_DATE_PARAMS
+#undef SHA_VAR_LEN_PARAMS
+
+// Add functions for decimal128
+GANDIVA_EXPORT
+const char* gdv_fn_sha256_decimal128(int64_t context, int64_t x_high, uint64_t x_low,
+                                     int32_t x_precision, int32_t x_scale,
```
Comment (emkornfield) [inline cpp/src/gandiva/gdv_function_stubs.cc:227]: precision and scale don't look like they are being used.  If these parameters are necessary to fit into gandivas framework.  Commenting them out as int32_t /*x_precision*/ would make these clearer.

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 13]
```diff
@@ -122,6 +123,133 @@ int32_t gdv_fn_populate_varlen_vector(int64_t context_ptr, int8_t* data_ptr,
...
+GANDIVA_EXPORT
+const char* gdv_fn_sha128_decimal128(int64_t context, int64_t x_high, uint64_t x_low,
+                                     int32_t x_precision, int32_t x_scale,
+                                     gdv_boolean x_isvalid, int32_t *out_length) {
+  if (!x_isvalid) {
+    *out_length = 0;
+    return "";
+  }
+
+  const auto &decimal_128 = gandiva::BasicDecimal128(x_high, x_low);
```
Comment (emkornfield) [inline cpp/src/gandiva/gdv_function_stubs.cc:248]: This seems like a strange formulation?  Why note BasicDecimal128 decimal_128(x_high, x_low);

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 14]
```diff
@@ -122,6 +123,133 @@ int32_t gdv_fn_populate_varlen_vector(int64_t context_ptr, int8_t* data_ptr,
   return 0;
 }
 
+#define SHA128_HASH_FUNCTION(TYPE)                                                    \
+  GANDIVA_EXPORT                                                                      \
+  const char *gdv_fn_sha128_##TYPE(int64_t context, gdv_##TYPE value,                 \
+                                        bool validity, int32_t *out_length) {         \
+    if (!validity) {                                                                  \
+      *out_length = 0;                                                                \
+      return "";                                                                      \
```
[the remark this answers is not in the document]
Comment (anthonylouisbsb) [reply cpp/src/gandiva/gdv_function_stubs.cc]: Added the implementation for the SHA hashes in null values.

File: cpp/src/gandiva/tests/hash_test.cc [hunk 15]
```diff
@@ -147,4 +153,252 @@ TEST_F(TestHash, TestBuf) {
...
+
+    auto response = outputs.at(0);
+    EXPECT_EQ(response->null_count(), 0);
+    EXPECT_EQ(response->GetScalar(0).ValueOrDie()->ToString(), "");
+    for (int i = 1; i < num_records; ++i) {
+        const auto &value_at_position = response->GetScalar(i).ValueOrDie()->ToString();
+        EXPECT_NE(value_at_position,
+				  response->GetScalar(i - 1).ValueOrDie()->ToString());
+    }
```
[the remark this answers is not in the document]
Comment (anthonylouisbsb) [reply cpp/src/gandiva/tests/hash_test.cc:443]: Added the asserts for the hash size in the integration tests.

File: cpp/src/gandiva/hash_utils.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (anthonylouisbsb) [reply cpp/src/gandiva/hash_utils.cc]: I fixed it.

File: cpp/src/gandiva/function_registry_common.h (hunk 1 above)
Comment (anthonylouisbsb) [reply cpp/src/gandiva/function_registry_common.h]: Already fixed.

File: cpp/src/gandiva/hash_utils.cc (hunk 2 above)
Comment (anthonylouisbsb) [reply cpp/src/gandiva/hash_utils.cc]: Fixed.

File: cpp/src/gandiva/hash_utils.cc (hunk 3 above)
Comment (anthonylouisbsb) [reply cpp/src/gandiva/hash_utils.cc:82]: Checks added.

File: cpp/src/gandiva/hash_utils.cc (hunk 4 above)
Comment (anthonylouisbsb) [reply cpp/src/gandiva/hash_utils.cc]: Suggestion applied.

File: cpp/src/gandiva/hash_utils.cc (hunk 8 above)
Comment (anthonylouisbsb) [reply cpp/src/gandiva/hash_utils.cc]: Added

File: cpp/src/gandiva/hash_utils.cc (hunk 7 above)
Comment (anthonylouisbsb) [reply cpp/src/gandiva/hash_utils.cc]: Added

File: cpp/src/gandiva/tests/hash_test.cc (hunk 6 above)
Comment (anthonylouisbsb) [reply cpp/src/gandiva/tests/hash_test.cc]: Changed the EXPECT_TRUE to ASSERT_OK

File: cpp/src/gandiva/hash_utils.h (hunk 5 above)
Comment (anthonylouisbsb) [reply cpp/src/gandiva/hash_utils.h]: I removed the class and changed the functions to first class and I added the documentation for the functions.

File: cpp/src/gandiva/hash_utils.cc (hunk 9 above)
Comment (anthonylouisbsb) [reply cpp/src/gandiva/hash_utils.cc]: Suggestion applied

File: cpp/src/gandiva/gdv_function_stubs.cc (hunk 13 above)
Comment (anthonylouisbsb) [reply cpp/src/gandiva/gdv_function_stubs.cc]: I changed to `const BasicDecimal128 decimal_128(x_high, x_low)`

File: cpp/src/gandiva/gdv_function_stubs.cc (hunk 12 above)
Comment (anthonylouisbsb) [reply cpp/src/gandiva/gdv_function_stubs.cc]: Suggestion applied

File: cpp/src/gandiva/gdv_function_stubs.cc (hunk 11 above)
Comment (anthonylouisbsb) [reply cpp/src/gandiva/gdv_function_stubs.cc:168]: I based in the implemented [hash functions inside the Gandiva](https://github.com/apache/arrow/blob/master/cpp/src/gandiva/precompiled/hash_test.cc) where the hashes for some values like `0` and `0.0` are the same, even if the types are differents.

File: cpp/src/gandiva/hash_utils.cc (hunk 4 above)
Comment (anthonylouisbsb) [reply cpp/src/gandiva/hash_utils.cc]: Suggestion applied

File: cpp/src/gandiva/hash_utils.cc (hunk 2 above)
Comment (anthonylouisbsb) [reply cpp/src/gandiva/hash_utils.cc]: Changed

File: cpp/src/gandiva/hash_utils.cc (hunk 3 above)
Comment (anthonylouisbsb) [reply cpp/src/gandiva/hash_utils.cc:91]: Checks added

File: cpp/src/gandiva/hash_utils.cc (hunk 10 above)
Comment (anthonylouisbsb) [reply cpp/src/gandiva/hash_utils.cc]: Format fixed

**Comment by anthonylouisbsb** [comment]
@emkornfield I applied all your suggestions and I left a comment I one implementation that you asked me https://github.com/apache/arrow/pull/9707#discussion_r597389422

File: cpp/src/gandiva/tests/hash_test.cc [hunk 16]
```diff
@@ -147,4 +153,252 @@ TEST_F(TestHash, TestBuf) {
   }
 }
 
+TEST_F(TestHash, TestSha256Simple) {
```
[the remark this answers is not in the document]
Comment (anthonylouisbsb) [reply cpp/src/gandiva/tests/hash_test.cc:156]: Tests added to the function.

File: cpp/src/gandiva/tests/hash_test.cc [hunk 17]
```diff
@@ -147,4 +149,271 @@ TEST_F(TestHash, TestBuf) {
...
+  // Evaluate expression
+  arrow::ArrayVector outputs;
+  status = projector->Evaluate(*in_batch, pool_, &outputs);
+  ASSERT_OK(status);
+
+  auto response = outputs.at(0);
+  const int sha1_hash_size = 40;
+
+  EXPECT_EQ(response->null_count(), 0);
+  for (int i = 1; i < num_records; ++i) {
```
Comment (projjal) [inline cpp/src/gandiva/tests/hash_test.cc:412]: can you also add assertion for the first record which is null. test case for hash of null value is missing.

Comment (anthonylouisbsb) [reply cpp/src/gandiva/tests/hash_test.cc:424]: Change applied. I added an assert to ensure that the SHA hash is not empty when the value is null and the hash size is equaled to the SHA size response.

**Review by projjal — APPROVED** [review]
Edit the PR title and description from SHA128 to SHA1

**Comment by anthonylouisbsb** [comment]
> Edit the PR title and description from SHA128 to SHA1

Already changed

**Review by projjal — APPROVED** [review]

**Comment by anthonylouisbsb** [comment]
@projjal I made a correction to fix the broken nightly build, It was a missing include header in the CMakeLists.txt of the Gandiva project

File: cpp/cmake_modules/FindOpenSSLAlt.cmake [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+find_package(OpenSSL)
+
+if (OPENSSL_FOUND)
+    if (OPENSSL_VERSION LESS "1.1.0")
+        message(SEND_ERROR "The OpenSSL must be grater than 1.0.0, the actual version is ${OPENSSL_VERSION}")
```
Comment (projjal) [inline cpp/cmake_modules/FindOpenSSLAlt.cmake:22]: >> grater than 1.0.0

greater than 1.1.0

File: dev/tasks/gandiva-jars/build-java.sh [hunk 19]
```diff
@@ -25,11 +25,11 @@ pushd java
   if [[ $OS_NAME == "linux" ]]; then
     SO_DEP=ldd
     GANDIVA_LIB="$CPP_BUILD_DIR"libgandiva_jni.so
-    WHITELIST=(linux-vdso libz librt libdl libpthread libstdc++ libm libgcc_s libc ld-linux-x86-64)
+    WHITELIST=(linux-vdso libz librt libdl libpthread libstdc++ libm libgcc_s libc ld-linux-x86-64 libssl libcrypto)
```
Comment (projjal) [inline dev/tasks/gandiva-jars/build-java.sh:28]: can you make sure that the nightly build uses a static version of openssl as dependency. this whitelist is to make sure that there are no shared libraries added as dependencies when building the gandiva library in nightly

Comment (anthonylouisbsb) [reply dev/tasks/gandiva-jars/build-java.sh]: Changes applied, the **ARROW_OPENSSL_USE_SHARED** variable will set if the library must be linked statically or dynamically.

File: cpp/cmake_modules/FindOpenSSLAlt.cmake (hunk 18 above)
Comment (projjal) [reply cpp/cmake_modules/FindOpenSSLAlt.cmake]: I meant greater than "1.1.0" (or rather greater than or equal to 1.1.0 to be correct)

Comment (anthonylouisbsb) [reply cpp/cmake_modules/FindOpenSSLAlt.cmake]: I changed now, check if it is correct

**Review by projjal — APPROVED** [review]

**Review by projjal — APPROVED** [review]

**Pull request closed without merging**